pallets/common/src/lib.rs

```diff
@@ -1295,12 +1300,18 @@ impl<T: Config> Pallet<T> {
 					//TODO: investigate threats during public minting.
 					let is_token_create =
 						is_token_create && (collection_admin || token_owner) && value.is_some();
-					if !(is_token_create
-						|| (collection_admin && is_collection_admin)
-						|| (token_owner && is_token_owner()?))
-					{
+					let partial_permission =
+						is_token_create || (collection_admin && is_collection_admin);
+
+					if !(partial_permission || (token_owner && is_token_owner()?)) {
 						fail!(<Error<T>>::NoPermission);
 					}
+
+					let has_token_owner = !partial_permission && token_owner;
+					let need_check_token_exist = !(has_token_owner && is_token_owner()?);
+					if need_check_token_exist && !is_token_exist() {
```

Comment:
Мне кажется, сейчас сложно понять как оно работает)
Много логических операций.

По сути, проверка на существование нужна только в двух случаях: при minting'е (public/admin) и когда админ пытается задать проперти.
Кроме того, я бы сделал проверку NoPermission более выразительной.

Этот код надо бы сделать максимально понятным, иначе, боюсь, еще чего пропустим :/

Критика предложенного кода приветствуется!
 
```rust
let is_valid_token_create = is_token_create && (collection_admin || token_owner) && value.is_some();
let admin_role_match = collection_admin && is_collection_admin;
let owner_role_match = || -> Result<bool, DispatchError> { token_owner && is_token_owner()?; }

if !is_valid_token_create && !admin_role_match && !owner_role_match()? {
    fail!(<Error<T>>::NoPermission);
}

if (!token_owner) {
   // Token owner role is not requested, the token existence is not checked.
   // However, we have permission to set up the property since we are inside this branch.

   if (!is_token_exist()) {
      fail!(<Error<T>>::TokenNotFound);
   }
}
```

Reply:
Поддерживаю, сейчас в это долго вдаваться и находить, что из чего течёт. Обфускация какая-то) Меня также смущают названия переменных. is_token_create (благо, есть в документации), is_token_exist лучше переименовать в is_token_being_created и token_exists.

Reply:
Насчет `is_token_being_created` плюсую. А вот насчет `token_exists` -- не уверен, присоединюсь к мнению большинства.

Reply:
is_token_exist кажется лучше, потому что token_exists звучит как переменная состояния, которую могут обновить, а не как проверка текущего состояния.

Reply:
token_owner_result.is_none() не получиться использовать потому что он имеет мутабельную ссылку в замыкании.

Reply:
Ну или использовать `Rc` для `token_owner_result`